gui/cache.go

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2020 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package gui
+
+import (
+	"math/big"
+	"sync"
+
+	"github.com/decred/dcrpool/pool"
+)
+
+// client represents a mining client. It is json annotated so it can easily be
+// encoded and sent over a websocket or pagination request.
+type client struct {
+	Miner    string `json:"miner"`
+	IP       string `json:"ip"`
+	HashRate string `json:"hashrate"`
+}
+
+// minedWork represents a block mined by the pool. It is json annotated so it
+// can easily be encoded and sent over a websocket or pagination request.
+type minedWork struct {
+	BlockHeight uint32 `json:"blockheight"`
+	BlockURL    string `json:"blockurl"`
+	MinedBy     string `json:"minedby"`
+	Miner       string `json:"miner"`
+	Confirmed   bool   `json:"confirmed"`
+	// AccountID holds the full ID (not truncated) and so should not be json encoded
+	AccountID string `json:"-"`
+}
+
+// workQuota represents dividend garnered by pool accounts through work
+// contributed. It is json annotated so it can easily be encoded and sent over a
+// websocket or pagination request.
+type workQuota struct {
+	AccountID string `json:"accountid"`
+	Percent   string `json:"percent"`
+}
+
+// Cache stores data which is required for the GUI. Each field has a setter and
+// getter, and they are protected by a mutex so they are safe for concurrent
+// access. Data returned by the getters is already formatted for display in the
+// GUI, so the formatting does not need to be repeated.
+type Cache struct {
+	blockExplorerURL string
+
+	minedWork    []minedWork
+	minedWorkMtx sync.RWMutex
+
+	workQuotas    []workQuota
+	workQuotasMtx sync.RWMutex
+
+	poolHash    string
+	poolHashMtx sync.RWMutex
+
+	clients    map[string][]client
+	clientsMtx sync.RWMutex
```

Comment:
Don't think it's necessary to group struct fields like this, I think th code base only has atomically updated fields separated from the rest.

Reply:
You mean the new lines? I included those just for the sake of readability and making it easier to see that each field has its own mutex, but I am happy to remove them.